libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,713 @@
+        // Negative return indicates error other than eof
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "write fail");
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+
+    response_payload->size = sizeof(uint32_t);
+    memcpy(response_payload->data, &bytes_written, request_payload->size);
+
+    return kErrNone;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Terminate(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    if (request_payload->session != 0 || _session_info.fd < 0) {
+        return kErrInvalidSession;
+    }
+
+    ::close(_session_info.fd);
+    _session_info.fd = -1;
+    _session_info.stream_download = false;
+
+    response_payload->size = 0;
+
+    return kErrNone;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Reset(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    if (_session_info.fd != -1) {
+        ::close(_session_info.fd);
+        _session_info.fd = -1;
+        _session_info.stream_download = false;
+    }
+
+    response_payload->size = 0;
+    return kErrNone;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_RemoveDirectory(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+    if (rmdir(_ftp_current_abs_path) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_CreateDirectory(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    if (mkdir(_ftp_current_abs_path, S_IRWXU | S_IRWXG | S_IRWXO) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_RemoveFile(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    if (unlink(_ftp_current_abs_path) == 0) {
+        response_payload->size = 0;
+        return kErrNone;
+    }
+    else {
+        _ftp_errno = errno;
+        return kErrFailErrno;
+    }
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_TruncateFile(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    _set_current_abs_path(_data_as_cstring(request_payload));
+
+    int ret = truncate(_ftp_current_abs_path, request_payload->offset);
+
+    if (ret == 0) {
+        return kErrNone;
+    }
+    _ftp_errno = errno;
+    return kErrFailErrno;
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_Rename(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    char *ptr = _data_as_cstring(request_payload);
+    size_t oldpath_sz = strlen(ptr);
+
+    if (oldpath_sz <= request_payload->size) {
```

Comment:
assumption 
size = oldpath 0x0 newpath 0x0
so the logic is wrong here should be >=